Add contributing document (#65)
chaiNNer-org/spandrel: CONTRIBUTING.md

The first step is to copy the code of the architecture into Spandrel.

Of course, the copy this code, we first have to find it! Unfortunately, project layout is not consistent across architecture repos, so you might be to search a bit. You need that file with the class that defines the architecture. In most repos, this file is called `model.py`, `models/<arch name>.py`, or `basicsr/archs/<arch name>_arch.py`. You know that you found the right file when it contains a class that is named after the architecture and inherits from `nn.Module`.

In the case of DITN, the file is called [`models/DITN_Real.py`](https://github.com/yongliuy/DITN/blob/3438e429c0538ee5061a7cfca587df0c4097703f/models/DITN_Real.py#L197).

Once you have found the file, create a new directory `src/spandrel/architectures/<arch name>/arch/` and copy the file into it. This directory will contain all code that we will copy. Since we respect the copy right of the original authors, we will also copy the `LICENSE` file of the repo into this directory.

In the case of DITN, we copy `models/DITN_Real.py` to `src/spandrel/architectures/DITN/arch/DITN_Real.py` and add the `LICENSE` file of the repo.

> The main model file might also reference other files in the repo. You have to copy those 2 into your `arch/` directory as well.

Reviewer: Which 2 files are you referring to?

In the case of DITN, the main model file doesn't reference any other files, so we are done.

### Step 1.1: Linting, type errors, and cleaning up

You might have already noticed that the code you copied lights up like a Christmas tree in your IDE. In VSCode, just hit Ctrl+S to let ruff fix the formatting and most linting errors.

However, there are a few things that we have to fix manually.

> - "Use `super()` instead of `super(__class__, self)`" \

Reviewer: Isn't there a ruff rule we can enable to fix this automatically?

Author: The rule doesn't auto fix, it only offers an IDE suggestion. That's why I mentioned this.

Finally, some model files contain a `if __name__ == "__main__":` section to run the model as a script. We do not ever want to run the model as a script, so we can just remove this section.

In the case of DITN, we just ignore a few type errors and remove [this section](https://github.com/yongliuy/DITN/blob/3438e429c0538ee5061a7cfca587df0c4097703f/models/DITN_Real.py#L249).

### Step 1.2: Fixing dependencies

The model files might have dependencies to external packages. We generally try to keep the number of dependencies for spandrel as low as possible, so we want to remove any dependencies that are not strictly necessary.

We allow models to have the following external dependencies: `torch`, `torchvision`, `numpy`, and `einops`.

`timm` is a special dependency for us, because we have vendored the most important code from `timm`. So instead of using the `timm` package, we should use the vendored code from `src/spandrel/architectures/__arch_helpers/timm/`.

In the case of DITN, its model file has no external dependencies, except for `torch` and `einops`, so we are done.

### Step 2: Adding a `load` method

> With the architecture code in place, we can start integrating it into spandrel. The first step is to define a `load` function. This function will take a state dict, detect the hyperparameters of the architecture (=the input arguments of the architecture's Python class), and return a loaded model.

Reviewer: Is that = a typo or intentional?

Author: It's intentional. I'm explaining what hyperparameters are.

    )
```

Next, fill in the template:

- Replace `architecture="ARCH_NAME"` with the name of your architecture.
- Replace the references to `ARCH_NAME` with your architecture model file and class.
- Replace `param1` and `param2` with the parameters of your architecture. Copy the default values of the parameters from the model class and pass them into the constructor (`model = ARCH_NAME(...)`).
- Fill in the missing value for `scale`, `input_channels`, and `output_channels`.
  - `scale`: If your architecture is a super resolution architecture, there should be a `scale`, `upscale`, or `upscale_factor` parameter. Use this value. If your architecture is not a super resolution architecture, set this value to `1`.
  - `input_channels`: The number of input channels of the architecture. Again, there should be a parameter for this. Something like `in_nc`, `in_channels`, `inp_channels`. If there isn't leave it at 3 for RGB images.
  - `output_channels`: The number of output channels of the architecture. Same as for `input_channels`. However, if there is no parameter for output channels, but there is one for input channels, just use the parameter for input channels.

> - Set the purpose of the architecture. If you have a restoration architecture (e.g. denoising, deblurring, JPEG), set this to "Restoration". If you have a super resolution architecture, set this to "SR" if the scale of the model is >1 and "Restoration" otherwise. We consider 1x model for SR architectures (e.g. 1x ESRGAN model) to be restoration models. (See DITN example below.)

Reviewer: ```suggestion
- Set the purpose of the architecture. If you have a restoration architecture (e.g. denoising, deblurring, or deJPEGing), set this to "Restoration". If you have a super resolution architecture, set this to "SR" if the scale of the model is >1 and "Restoration" otherwise. We consider 1x model for SR architectures (e.g. 1x ESRGAN model) to be restoration models. (See DITN example below.)
```

Author: > deJPEGing

I hate this. Let's got with "removing JPEG artifacts".

If you [run this test](#running-tests) now, it should pass. Since we are passing in the default values for all parameters in `load`, we should get the same model as just `ARCH_NAME()`.

With this now all setup up, we can start detecting parameters in the next step.

But before that, here's how this test looks like for DITN:

<details>
<summary>Click to see code</summary>

```python
from spandrel.architectures.DITN import DITN, load

from .util import assert_loads_correctly

def test_DITN_load():
    assert_loads_correctly(
        load,
        lambda: DITN(),
        condition=lambda a, b: (
            a.patch_size == b.patch_size
            and a.dim == b.dim
            and a.scale == b.scale
            and a.SAL_blocks == b.SAL_blocks
            and a.ITL_blocks == b.ITL_blocks
        ),
    )
```

> Notices how `condition` does not cover all parameters. This particular model doesn't store all parameters as fields, so we don't compare them here.

Reviewer: ```suggestion
Notice how `condition` does not cover all parameters. This particular model doesn't store all parameters as fields, so we don't compare them here.
```

Open `src\spandrel\__helpers\main_registry.py` and start by importing your architecture in the `from ..architectures import (...)` state. Then get read to scroll all the way down and add a new `ArchSupport` object at the end of `MAIN_REGISTRY.add(...)`. The `ArchSupport` object will tell spandrel how to detect whether an arbitrary state dict (so a state dict of any architecture) is from your architecture. We do this by simply detecting the presence of a particular key. Each architecture has a few keys that are always present in the state dict, and we'll use those for detection.

In the case of DITN, its `ArchSupport` object looks like this:

```python
ArchSupport(
    id="DITN",
    detect=_has_keys(
        "sft.weight",
        "UFONE.0.ITLs.0.attn.temperature",
        "UFONE.0.ITLs.0.ffn.project_in.weight",
        "UFONE.0.ITLs.0.ffn.dwconv.weight",
        "UFONE.0.ITLs.0.ffn.project_out.weight",
        "conv_after_body.weight",
        "upsample.0.weight",
    ),
    load=DITN.load,
)
```

> Run the test from before again, it should now fail because there's not snapshot for it yet. So let's run the architecture tests from the command line to fix this: `pytest tests/test_ARCH.py --snapshot-update`. Running this, all tests should pass, and a new snapshot should be created. You'll also see a few now images. Those are the snapshots for our inference tests.

Reviewer: ```suggestion
Run the test from before again, it should now fail because there's not snapshot for it yet. So let's run the architecture tests from the command line to fix this: `pytest tests/test_ARCH.py --snapshot-update`. Running this, all tests should pass, and a new snapshot should be created. You'll also see a few new images. Those are the snapshots for our inference tests.
```
